cps/main.c

```diff
@@ -672,6 +679,73 @@ fill_bgp_rule(struct ipv6_acl_rule *rule, struct gatekeeper_if *iface,
 	}
 }
 
+/*
+ * Match the packet if it fails to be classifed by ACL rules,
+ * if it's bgp packet, then submit it to the LLS block.
+ *
+ * Return values: -ENOENT for no matching, -EINVAL for submit_bgp error,
+ * and 0 for successful match and submit.
+ */
+static int
+match_bgp(struct rte_mbuf *pkt, struct gatekeeper_if *iface)
+{
+	/*
+	 * The TCP header offset in terms of the
+	 * beginning of the IPv6 header.
+	 */
+	int tcp_offset;
+	uint8_t nexthdr;
+	const uint16_t BE_ETHER_TYPE_IPv6 = rte_cpu_to_be_16(ETHER_TYPE_IPv6);
+	struct ether_hdr *eth_hdr =
+		rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	struct ipv6_hdr *ip6hdr;
+	struct tcp_hdr *tcp_hdr;
+	uint16_t minimum_size = sizeof(*eth_hdr) +
+		sizeof(struct ipv6_hdr) + sizeof(struct tcp_hdr);
+
+	if (unlikely(eth_hdr->ether_type != BE_ETHER_TYPE_IPv6))
+		return -ENOENT;
+
+	if (pkt->data_len < minimum_size) {
+		RTE_LOG(NOTICE, GATEKEEPER, "cps: BGP packet received is %"PRIx16" bytes but should be at least %hu bytes\n",
+			pkt->data_len, minimum_size);
+		return -ENOENT;
+	}
+
+ 	ip6hdr = (struct ipv6_hdr *)&eth_hdr[1];
+
+	if ((memcmp(ip6hdr->dst_addr, &iface->ip6_addr,
+			sizeof(iface->ip6_addr)) != 0) &&
+			(memcmp(ip6hdr->dst_addr, &iface->ll_ip6_addr,
+			sizeof(iface->ll_ip6_addr)) != 0) &&
+			(memcmp(ip6hdr->dst_addr, &iface->ip6_mc_addr,
+			sizeof(iface->ip6_mc_addr)) != 0) &&
+			(memcmp(ip6hdr->dst_addr,
+			&iface->ll_ip6_mc_addr,
+			sizeof(iface->ll_ip6_mc_addr)) != 0))
+		return -ENOENT;
+
+	nexthdr = ip6hdr->proto;
```

Comment:
This line is no longer needed since `ipv6_skip_exthdr()` properly initializes `nexthdr`.